controllers/drplacementcontrol_controller.go

```diff
@@ -953,56 +1024,153 @@ func (d *DRPCInstance) switchToPreferredCluster(drState rmn.DRState) (bool, erro
 	d.setDRState(drState)
 	setMetricsTimerFromDRState(drState, timerStart)
 
-	// During failback or relocation, both clusters should be up and both must be secondaries before we proceed
-	ensured, err := d.processVRGForSecondaries()
-	if err != nil || !ensured {
-		return !done, err
+	// During failback/relocate, the preferredCluster is does not contain a VRG or the VRG is already
+	// secondary. We need to skip checking if the VRG for it is secondary to avoid messing up with the
+	// order of execution (it could be refactored better to avoid this complexity). IOW, if we first update
+	// VRG in all clusters to secondaries, then we call runPlacementTask. If runPlacementTask does not complete
+	// in one shot, then coming back to this loop will reset the preferredCluster to secondary again.
+	clusterToSkip := preferredCluster
+	if !d.ensureVRGIsSecondary(clusterToSkip) {
+		// During failback or relocation, both clusters should be up and both must be secondaries before we proceed.
+		ensured, err := d.processVRGForSecondaries()
+		if err != nil || !ensured {
+			return !done, err
+		}
 	}
 
-	result, err := d.runPlacementTask(preferredCluster, preferredClusterNamespace)
+	const restorePVs = true
+
+	result, err := d.runPlacementTask(preferredCluster, preferredClusterNamespace, restorePVs)
 	if err != nil {
 		return !done, err
 	}
 
-	// 8. All good so far, update DRPC decision and state
+	// All good so far, update DRPC decision and state
 	if len(d.userPlacementRule.Status.Decisions) > 0 {
 		d.instance.Status.PreferredDecision = d.userPlacementRule.Status.Decisions[0]
 	}
 
 	d.advanceToNextDRState()
-	d.log.Info("Done", "Last known state", d.instance.Status.LastKnownDRState)
+	d.log.Info("Done", "Last known state", d.getLastDRState())
 	setMetricsTimerFromDRState(drState, timerStop)
 
 	return result, nil
 }
 
-// runPlacementTast is a series of steps to creating, updating, and cleaning up
+// runPlacementTask is a series of steps to creating, updating, and cleaning up
 // the necessary objects for the failover, failback, or relocation
-func (d *DRPCInstance) runPlacementTask(targetCluster, targetClusterNamespace string) (bool, error) {
-	// Restore PV to preferredCluster
-	err := d.createPVManifestWorkForRestore(targetCluster)
+//nolint:cyclop
+func (d *DRPCInstance) runPlacementTask(targetCluster, targetClusterNamespace string, restorePVs bool) (bool, error) {
+	d.log.Info("runPlacementTask", "cluster", targetCluster, "restorePVs", restorePVs)
+
+	created, err := d.createManifestWorks(targetCluster)
 	if err != nil {
 		return false, err
 	}
 
-	err = d.updateUserPlRuleAndCreateVRGMW(targetCluster, targetClusterNamespace)
+	if created && restorePVs {
+		// We just created MWs. Give it time until the PV restore is complete
+		return false, fmt.Errorf("%w)", WaitForPVRestoreToComplete)
+	}
+
+	if !created {
+		updated, err := d.updateManifestWorkToPrimary(targetCluster)
```

Comment:
This is dangerous! We possibly need to address this in the next PR.

VRG cannot shift from Secondary to Primary, as it will hold back the PVC and hence the PV till deleted. Thus, if we do the shift, either the PV restore would fail or PVCs would get new PVs instead of the existing restored one.

Currently, while testing relocate we ensure VRG is deleted on all other clusters (manually) and hence it would never get created or left behind as Secondary avoiding the problem. BUT we also never shifted the state from Secondary to Primary, which we should ensure.

This adds more complexity to the already noted code comment above in line 1027.

Reply:
hmmm. So this goes along with my previous comment. We ensure that the VRG on the peer has transitioned to secondary before we get here.  Again the question, isn't that enough?